shared/src/main/java/com/couchbase/lite/LiveQuery.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    //---------------------------------------------
+    // Implementation of DatabaseChangeListener
+    //---------------------------------------------
+    @Override
+    public void changed(DatabaseChange change) {
+        if (willUpdate)
+            return; // Already a pending update scheduled
+
+        // Schedule an update, respecting the updateInterval:
+        long updateDelay = lastUpdatedAt + kDefaultLiveQueryUpdateInterval - System.currentTimeMillis();
+        updateDelay = Math.max(0, Math.min(this.kDefaultLiveQueryUpdateInterval, updateDelay));
+        update(updateDelay);
+    }
+
+    //---------------------------------------------
+    // Package level access
+    //---------------------------------------------
+
+    /**
+     * The current query results; this updates as the database changes.
+     * Its value will be null until the initial asynchronous query finishes.
+     * <p>
+     * NOTE: for unit test
+     */
+    /*package*/ ResultSet getResultSet() {
+        run();
+        return resultSet;
+    }
+
+    //---------------------------------------------
+    // Private (in class only)
+    //---------------------------------------------
+
+    /**
+     * @param delay millisecond
+     */
+    private void update(long delay) {
+        if (willUpdate)
+            return; // Already a pending update scheduled
+
+        willUpdate = true;
+        // create one doc
+        new Handler(Looper.getMainLooper())
+                .postDelayed(new Runnable() {
+                    @Override
+                    public void run() {
+                        update();
+                    }
+                }, delay); // ms
+    }
+
+    private void update() {
+        // TODO: Make this asynchronous (as in 1.x)
+        try {
+            Log.i(TAG, "%s: Querying...", this);
+            ResultSet oldResultSet = resultSet;
+            ResultSet newResultSet;
+            if (oldResultSet == null || !oldResultSet.isValidEnumerator() || forceReload)
+                newResultSet = query.run();
+            else
+                newResultSet = oldResultSet.refresh();
+
+            willUpdate = forceReload = false;
+            lastUpdatedAt = System.currentTimeMillis();
+
+            if (newResultSet != null) {
+                if (oldResultSet != null)
+                    Log.i(TAG, "%s: Changed!", this);
+                resultSet = newResultSet;
+                sendNotification(new LiveQueryChange(this, resultSet, null));
+            } else {
+                Log.i(TAG, "%s: ...no change", this);
+            }
+        } catch (CouchbaseLiteException e) {
+            sendNotification(new LiveQueryChange(null, e));
+        }
+    }
+
+    private void sendNotification(LiveQueryChange change) {
+        synchronized (queryChangeListener) {
+            for (LiveQueryChangeListener listener : queryChangeListener) {
+                listener.changed(change);
+            }
+        }
+    }
+}
```

Comment:
Should also remove the listener to the database when the object is collected as well.